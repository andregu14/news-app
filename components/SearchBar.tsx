import React, { useState } from "react";
import {
  TextInput,
  StyleSheet,
  useColorScheme,
  TouchableOpacity,
} from "react-native";
import { View, ViewProps } from "./Themed";
import { MaterialCommunityIcons } from "@expo/vector-icons";
import Colors from "../constants/Colors";

type SearchBarProps = ViewProps & {
  onSubmitEditing?: (query: string) => void;
  placeholder?: string;
  defaultValue?: string;
};

export default function SearchBar({
  onSubmitEditing,
  placeholder = "Pesquisar",
  defaultValue = "",
  style,
  ...otherProps
}: SearchBarProps) {
  const colorScheme = useColorScheme() ?? "light";
  const themeColors = Colors[colorScheme];
  const [inputValue, setInputValue] = useState(defaultValue);

  const handleClear = () => {
    setInputValue("");
  };

  const handleSubmit = () => {
    onSubmitEditing?.(inputValue); // Chamar onSubmitEditing com o valor interno
    setInputValue("");
  };

  return (
    <View
      style={[
        styles.container,
        {
          borderColor:
            themeColors.tint === Colors.dark.tint
              ? "rgba(255, 255, 255, 0.2)"
              : "#eee",
        },
        style,
      ]}
      darkColor="rgba(255, 255, 255, 0.10)"
      lightColor="#fff"
      {...otherProps}
    >
      {/* ... ícone ... */}
      <View
        style={[
          styles.iconContainer,
          {
            backgroundColor:
              themeColors.tint === Colors.dark.tint ? "#3F3F3F" : "#eee",
          },
        ]}
      >
        <MaterialCommunityIcons
          name="magnify"
          size={28}
          color={themeColors.tint}
        />
      </View>
      <TextInput
        value={inputValue}
        onChangeText={setInputValue}
        onSubmitEditing={handleSubmit}
        placeholder={placeholder}
        placeholderTextColor={
          themeColors.tint === Colors.dark.tint
            ? "rgba(255, 255, 255, 0.5)"
            : "rgba(0, 0, 0, 0.5)"
        }
        style={[
          styles.input,
          {
            borderColor:
              themeColors.tint === Colors.dark.tint
                ? "rgba(255, 255, 255, 0.2)"
                : "#eee",
            color: themeColors.text,
          },
        ]}
        inputMode="search"
        returnKeyType="search"
        accessibilityLabel="Campo de pesquisa"
        accessibilityHint="Digite aqui para pesquisar notícias"
      />
      {/* Botão de limpar aparece se houver texto */}
      {inputValue.length > 0 && (
        <TouchableOpacity
          onPress={handleClear}
          style={styles.clearButton}
          accessibilityLabel="Limpar pesquisa"
        >
          <MaterialCommunityIcons
            name="close-circle"
            size={20}
            color={
              themeColors.tint === Colors.dark.tint
                ? "rgba(255, 255, 255, 0.5)"
                : "rgba(0, 0, 0, 0.5)"
            }
          />
        </TouchableOpacity>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    height: 50,
    borderWidth: 1,
    borderRadius: 14,
    flexDirection: "row",
    alignItems: "center",
  },
  iconContainer: {
    height: "100%",
    width: 50,
    justifyContent: "center",
    alignItems: "center",
    borderTopLeftRadius: 13,
    borderBottomLeftRadius: 13,
  },
  input: {
    height: "100%",
    flex: 1,
    paddingLeft: 15,
    paddingRight: 35,
    borderLeftWidth: 1,
    fontSize: 16,
  },
  clearButton: {
    position: "absolute",
    right: 10,
    height: "100%",
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: 5,
  },
});
