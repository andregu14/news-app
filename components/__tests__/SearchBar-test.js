import { render, screen, fireEvent } from '@testing-library/react-native';
import SearchBar from '../SearchBar';

describe('<SearchBar />', () => {
    it('renders correctly and handles input', () => {
        const onSubmitEditing = jest.fn();
        render(<SearchBar placeholder="Search..." onSubmitEditing={onSubmitEditing} />);
        const input = screen.getByPlaceholderText('Search...');
        expect(input).toBeVisible();
        fireEvent.changeText(input, 'Test Input');
        fireEvent(input, 'submitEditing');
        expect(onSubmitEditing).toHaveBeenCalledWith('Test Input');
        expect(screen.toJSON()).toMatchSnapshot();
    });

    it('handles text submission correctly', () => {
        const onSubmitEditing = jest.fn();
        render(<SearchBar placeholder="Search..." onSubmitEditing={onSubmitEditing} />);
        const input = screen.getByPlaceholderText('Search...');
        fireEvent.changeText(input, 'Test Query');
        fireEvent(input, 'submitEditing');
        expect(onSubmitEditing).toHaveBeenCalledWith('Test Query');
    });

    it('clears input when clear button is pressed', () => {
        render(<SearchBar placeholder="Search..." />);
        const input = screen.getByPlaceholderText('Search...');
        fireEvent.changeText(input, 'Test Query');
        const clearButton = screen.getByLabelText('Limpar pesquisa');
        fireEvent.press(clearButton);
        expect(input.props.value).toBe('');
    });
});
