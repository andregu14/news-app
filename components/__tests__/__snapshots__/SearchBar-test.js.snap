// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders correctly and handles input 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      [
        {
          "alignItems": "center",
          "borderRadius": 14,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 50,
        },
        {
          "borderColor": "#eee",
        },
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "alignItems": "center",
            "borderBottomLeftRadius": 13,
            "borderTopLeftRadius": 13,
            "height": "100%",
            "justifyContent": "center",
            "width": 50,
          },
          {
            "backgroundColor": "#eee",
          },
        ],
      ]
    }
  >
    <Text />
  </View>
  <TextInput
    accessibilityHint="Digite aqui para pesquisar notÃ­cias"
    accessibilityLabel="Campo de pesquisa"
    inputMode="search"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search..."
    placeholderTextColor="rgba(0, 0, 0, 0.5)"
    returnKeyType="search"
    style={
      [
        {
          "borderLeftWidth": 1,
          "flex": 1,
          "fontFamily": "Inter_400Regular",
          "fontSize": 16,
          "height": "100%",
          "paddingLeft": 15,
          "paddingRight": 35,
        },
        {
          "borderColor": "#eee",
          "color": "#000",
        },
      ]
    }
    value=""
  />
</View>
`;
